### PR TITLE
Add support for failureDomain (AZ) for bastion

### LINKED ISSUE #753: Support failureDomain (availability zone) for bastion
/kind feature

**Describe the solution you'd like**

Today it's possible to set `failureDomain` on OpenStack machines (via `MachineDeployment`).

We'd like to be able to set this for the **bastion** too.

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #753 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jichenjc** [comment]
@MPV let me know your thoughts, thanks

File: api/v1alpha3/types.go
Comment (sbueringer) [inline api/v1alpha3/types.go:267]: Is this really a failuredomain in the generic sense that we have it in the KubeadmControlPlane?

I would have guessed we have only one bastion host and we simply specific the availabilty zone here in which it should be created.

**Review by sbueringer — CHANGES_REQUESTED** [review]

**Comment by sbueringer** [comment]
@jichenjc @hidekazuna I would really like to merge #748 relatively soon. I have to rebase / re-implement all those features again on my branch :/

File: api/v1alpha3/types.go
Comment (MPV) [reply api/v1alpha3/types.go:267]: Looks to me like you're right. But when creating an Instance (like both Bastion and control plane nodes seem to be doing similarly), they just translate the FailureDomain from the CAPx spec into OpenStack AZs here?

Example:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/30ddae05f68056feb692318bb8c3479869eee034/pkg/cloud/services/compute/instance.go#L228

So I guess this looks fine to me. 👍

**Comment by hidekazuna** [comment]
@sbueringer You can announce code freeze for merging #748 in slack channel :)

**Comment by jichenjc** [comment]
/hold

per @sbueringer ,let's focus on 748 first and after that, I will do the rebase work

**Comment by sbueringer** [comment]
> @sbueringer You can announce code freeze for merging #748 in slack channel :)

Oh thx for the hint. Not sure how I missed that one :) Was thinking about what the best way is to announce that.

**Comment by sbueringer** [comment]
> /hold
> 
> per @sbueringer ,let's focus on 748 first and after that, I will do the rebase work

Thx, I also rebased my branch to the current master.

**Comment by sbueringer** [comment]
@jichenjc As release-0.3 has been created with the 0.3.4 release we should probably cherry-pick all merges after the 0.3.4 release on the release branch if someone wants them in 0.3.x

File: api/v1alpha3/types.go
Comment (sbueringer) [reply api/v1alpha3/types.go:259]: I'm not entirely sure but I think we only do this for control plane machines. For all other machines it's just setting the az on a OpenStackMachineSpec

FailureDomain as a concept is meant to abstract away from Specific IaaS provider because failure domains can be very differently (especially regarding hierarchy) between different cloud providers.

Comment (sbueringer) [reply api/v1alpha3/types.go:259]: Just forget everything I've said :)

You're absolutely right, we're just working with FailureDomains everywhere. Not sure how I got the idea that we specify az's on any resources. Probably mixed it up with our internal proprietar non-ClusterAPI implementation :)

File: api/v1alpha3/types.go
Comment (sbueringer) [inline api/v1alpha3/types.go:258]:
```suggestion
	//+optional
```

File: api/v1alpha3/types.go
Comment (sbueringer) [reply api/v1alpha3/types.go:259]: Okay so overall looks good to me. Although I suspect before going to beta someday we should do a full API redesign (not especially because of this PR but our resources could use a well thought out design after the history we have behind us :)).

Comment (MPV) [reply api/v1alpha3/types.go:259]: Well, I guess you're right too? 😊 

`OpenStackCluster` has `spec.controlPlaneAvailabilityZones`
`MachineDeployment` has `spec.template.spec.failureDomain`

Comment (sbueringer) [reply api/v1alpha3/types.go:259]: We should probably start an issue with API inconsistencies which we can then all fix when we redesign it. I guess for v1alpha4 it's not worth the breaking change

File: api/v1alpha3/types.go
Comment (sbueringer) [inline api/v1alpha3/types.go:253]: Does it still work like this? I would have thought it hast to look like `//+optional`?

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
with this change, I can have 

```
+--------------------------------------+------------------------------------+--------+------------+-------------+-------------------------------------------------------------------------+
| ID                                   | Name                               | Status | Task State | Power State | Networks                                                                |
+--------------------------------------+------------------------------------+--------+------------+-------------+-------------------------------------------------------------------------+
| 17af2aea-e79f-4768-9fc4-b744045e11d3 | capi-openstack-3-bastion           | ACTIVE | -          | Running     | k8s-clusterapi-cluster-default-capi-openstack-3=10.6.0.178, 172.24.4.25 |
| b6348cff-4a47-4af2-b48a-ee0a03dd803b | capi-openstack-control-plane-j7mjg | ACTIVE | -          | Running     | k8s-clusterapi-cluster-default-capi-openstack-3=10.6.0.159, 172.24.4.2  |
| f37b3ee6-6b2e-4696-8627-882209c8f850 | capi-openstack-md-0-zjj5v          | ACTIVE | -          | Running     | k8s-clusterapi-cluster-default-capi-openstack-3=10.6.0.104              |
+--------------------------------------+------------------------------------+--------+------------+-------------+-------------------------------------------------------------------------+

root@jjtest13:~/abc# cinder list
+--------------------------------------+--------+------+------+-------------+----------+--------------------------------------+
| ID                                   | Status | Name | Size | Volume Type | Bootable | Attached to                          |
+--------------------------------------+--------+------+------+-------------+----------+--------------------------------------+
| bf7ff123-a5ee-4d18-9af0-11c433007993 | in-use |      | 1    | lvmdriver-1 | true     | 17af2aea-e79f-4768-9fc4-b744045e11d3 |
+--------------------------------------+--------+------+------+-------------+----------+--------------------------------------+

```

with 
```
  bastion:
    enabled: true
    failureDomain: nova
    instance:
      flavor: m1.tiny
      image: cirros-0.5.1-x86_64-disk
      rootVolume:
        diskSize: 1
        sourceType: "image"
        sourceUUID: d08b6b0f-361a-4586-860a-949b623abefa

```

**Review by sbueringer — APPROVED** [review]

**Comment by MPV** [comment]
Great job here! ❤️

> @jichenjc As release-0.3 has been created with the 0.3.4 release we should probably cherry-pick all merges after the 0.3.4 release on the release branch if someone wants them in 0.3.x

Cherry-picking this into a 0.3.x release sounds great. 😍

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
@hidekazuna can you help to review this ? thanks

File: api/v1alpha3/types.go
Comment (hidekazuna) [reply api/v1alpha3/types.go:260]: @jichenjc When I borrow bastion concept from CAPA, I too wondered why bastion does not use AWSMachineSpec. 
Bastion is not openStackMachine, but member of OpenStackClusterSpec. So I thought we should not use OpenStackMachineSpec. 
FailureDomain is somehow considered related to Cluster API. Bastion concept is not directly related to Cluster API. 

Simply I suggest to add `AvailabilityZone string`

Comment (jichenjc) [reply api/v1alpha3/types.go:260]: I understand , but I think we don't need fully follow AWS provider here? 
bastion to me it's `some special param` + `a VM` , so if we keep adding stuffs like root volume, security groups 
ssh keys, then finally it will became a MachineSpec any way..

I can move to AvailabilityZone

**Comment by hidekazuna** [comment]
OK.. 

/lgtm

**Comment by jichenjc** [comment]
@hidekazuna @sbueringer need your lgtm :)

**Comment by MPV** [comment]
Wonderful @jichenjc! 🎉

Any chance we're seeing this in a 0.03x release soon? 🙏🏻

**Comment by jichenjc** [comment]
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/770  is the backport

**Pull request merged**
